⬆️  Bump Gateway API version to v0.5.0-rc1

carlisia opened the pull request:
Closes #306.

- Upgrade Gateway API import paths + deps to v0.5.0-rc1
- Update CRDs, config, instructions + point to local CRDs
- ParentRef -> ParentReference
- ConditionRouteAccepted -> RouteConditionAccepted

## Todos

- Regenerate client code to address the build error below (or change it manually?). The `v0.5.0-rc1` changes the return type of `ReferencePolicyList` to a slice of the new `ReferenceGrant `:

https://pkg.go.dev/sigs.k8s.io/gateway-api@v0.5.0-rc1/apis/v1alpha2?utm_source=gopls#ReferencePolicyList
```
[⚡ PIPE] ✗  go build ./...
# knative.dev/net-gateway-api/pkg/client/gatewayapi/injection/informers/apis/v1alpha2/referencepolicy
pkg/client/gatewayapi/injection/informers/apis/v1alpha2/referencepolicy/referencepolicy.go:107:15: cannot use &lo.Items[idx] (type *"sigs.k8s.io/gateway-api/apis/v1alpha2".ReferenceGrant) as type *"sigs.k8s.io/gateway-api/apis/v1alpha2".ReferencePolicy in append
# knative.dev/net-gateway-api/pkg/client/gatewayapi/injection/informers/apis/v1alpha2/referencepolicy/filtered
pkg/client/gatewayapi/injection/informers/apis/v1alpha2/referencepolicy/filtered/referencepolicy.go:126:15: cannot use &lo.Items[idx] (type *"sigs.k8s.io/gateway-api/apis/v1alpha2".ReferenceGrant) as type *"sigs.k8s.io/gateway-api/apis/v1alpha2".ReferencePolicy in append
```

- Fix this other build error:
```
# knative.dev/pkg/client/injection/kube/client
vendor/knative.dev/pkg/client/injection/kube/client/client.go:13323:9: cannot use &wrapStorageV1{...} (type *wrapStorageV1) as type "k8s.io/client-go/kubernetes/typed/storage/v1".StorageV1Interface in return argument:
        *wrapStorageV1 does not implement "k8s.io/client-go/kubernetes/typed/storage/v1".StorageV1Interface (missing CSIStorageCapacities method)
```

carlisia commented:
/assign @nak3

dprotaso commented on README.md:
````diff
@@ -30,7 +37,7 @@ kubectl apply -f https://github.com/knative/serving/releases/latest/download/ser
 
 ```bash
-ko apply -f config/
+ko apply -f third_party/gateway-api/
````
Things specific to this repo should probably remain in `config/` 

ie. controller, cluster role, config

I'm indifferent whether the gateway CRDs are in `config/` or `third_party` - ie. in contour we keep things in a separate nested folder under `config`

Probably worth thinking about if we're going to ship the gateway CRDS with the net-gateway-api release

Any yamls we move around we should update the release script

https://github.com/knative-sandbox/net-gateway-api/blob/main/hack/release.sh

dprotaso commented on go.mod:
```diff
@@ -2,21 +2,28 @@ module knative.dev/net-gateway-api
 go 1.17
 
+replace (
+	k8s.io/api v0.28.3 => k8s.io/api v0.24.1
+	k8s.io/apimachinery v0.28.3 => k8s.io/apimachinery v0.24.1
+	k8s.io/client-go v0.28.3 => k8s.io/client-go v0.24.1
+	k8s.io/code-generator v0.28.3 => k8s.io/code-generator v0.24.1
```
nit: wrong version

carlisia commented:
/retest

evankanderson commented on test/e2e-common.sh:
```diff
@@ -41,7 +41,7 @@ function test_setup() {
   # Bringing up controllers.
   echo ">> Bringing up controller"
-  ko apply -f config/ || return 1
+  ko apply -f third_party/gateway-api/ || return 1
```
This doesn't look like it's bringing up the controllers anymore. I don't know if we need both applied. In particular, line 56 says "Deploy Gateway API resources".

carlisia replied:
Good catch wrt ln 56. That bit already is in the istio specific e2e test file (kind-e2e-istio.sh), as it should. I'll remove it.

evankanderson reviewed (commented):
FWIW, the `100-`, `200-`, etc prefixes on files in `/config` are pretty standard elsewhere in Knative (which makes it easier to put in more config files), but I don't have a strong feeling about `01-`, `03-` and `04-` instead.

There are a couple places where I think you ended up replacing `/config` with `/third_party/gateway-api` which you didn't mean to, but this otherwise looks as reasonable as can be expected given the changes between 0.4.0 and 0.5.0-rc1.

Thanks!

dprotaso commented:
Latest gateway-api @ HEAD has the fix (https://github.com/kubernetes-sigs/gateway-api/pull/1239) to unblock injection/code generation - doing a `go get sigs.k8s.io/gateway-api@main` should pull in the changes

carlisia commented:
I'm fixing the e2e test scripts so they (hopefully) each contain what's needed for starting the corresponding controller (and waiting for it to be ready), and no more.

evankanderson commented:
Let me know when this goes in. I think you wanted this before #316, so I'm going to hold that one since it's still open.

carlisia commented:
Argh, I am sorry that I only saw this now. Wanted to reorg the e2e tests in a separate PR but realized it needs to be done in a way that is compatible with how serving loads tests and artifacts, and so both sides need to be changed.

I will spruce up a couple minor things and have this PR ready tomorrow morning.

carlisia commented on hack/update-codegen.sh:
```diff
@@ -61,8 +65,13 @@ group "Update gateway API CRDs"
 then
-  echo "# Generated with \"kubectl kustomize https://github.com/kubernetes-sigs/gateway-api/config/crd?ref=${GATEWAY_API_VERSION}\"" > ${REPO_ROOT_DIR}/config/100-gateway-api.yaml
-  kubectl kustomize "https://github.com/kubernetes-sigs/gateway-api/config/crd?ref=${GATEWAY_API_VERSION}" >> ${REPO_ROOT_DIR}/config/100-gateway-api.yaml
+  echo "# Generated with \"kubectl kustomize github.com/kubernetes-sigs/gateway-api/config/crd/experimental?ref=${GATEWAY_API_VERSION}" > "${REPO_ROOT_DIR}/third_party/gateway-api/00-crds.yaml"
```
It works :)

But putting it back for consistency.

carlisia commented:
@evankanderson thanks for the nudge and for deferring #316 til this PR merges. All code reviews were addressed.

carlisia commented:
/test integration-tests_net-gateway-api_main

evankanderson commented:
/retest

nak3 commented:
The logs is complaining `"istio-ingressgateway" not found` as:

```    
    stream.go:254: E 17:28:17.621 net-gateway-api-controller-965c77dfb-m5hhc [knative.dev.net-gateway-api.pkg.reconciler.ingress.Reconciler] [serving-tests/ingress-conformance-websocket-uvaxcupm] Returned an error err=failed to probe Ingress: failed to get endpoints: endpoints "istio-ingressgateway" not found
    stream.go:254: E 17:28:17.622 net-gateway-api-controller-965c77dfb-m5hhc [knative.dev.net-gateway-api.pkg.reconciler.ingress.Reconciler] [serving-tests/ingress-conformance-websocket-uvaxcupm] Reconcile error err=failed to probe Ingress: failed to get endpoints: endpoints "istio-ingressgateway" not found
```

The istio was deployed via `test_setup()` before but this PR stops deploying it.

evankanderson commented:
It looks like (maybe in the added include?) we're attempting to change the value of a read-only variable:

    + ./test/presubmit-tests.sh --build-tests
       ./test/../vendor/knative.dev/hack/library.sh: line 28: KNATIVE_TESTS_PROJECT: readonly variable

carlisia commented:
> It looks like (maybe in the added include?) we're attempting to change the value of a read-only variable
Yes, I fixed that.

Trying to figure out what change I made that is causing this failure:

This failure:
```
>> prow-tests image version
Prow test image was built from 9ea214bff8613f44e2151255905c654f0619079f commit which is viewable at https://github.com/knative/test-infra/tree/9ea214bff8613f44e2151255905c654f0619079f 
===================================
==== RUNNING INTEGRATION TESTS ====
===================================
==== Tue Jul  5 22:04:58 UTC 2022
===================================
Running integration test test/e2e-tests.sh
The connection to the server localhost:8080 was refused - did you specify the right host or port?
>> Setting up logging...
>> Publishing test images
test/upload-test-images.sh: line 35: KO_DOCKER_REPO: You must set 'KO_DOCKER_REPO', see DEVELOPMENT.md
ERROR: Error uploading test images
```

...when the last successful run (`main`) shows:
```
>> prow-tests image version
Prow test image was built from 9ea214bff8613f44e2151255905c654f0619079f commit which is viewable at https://github.com/knative/test-infra/tree/9ea214bff8613f44e2151255905c654f0619079f 
===================================
==== RUNNING INTEGRATION TESTS ====
===================================
==== Mon Jul  4 14:20:30 UTC 2022
===================================
Running integration test test/e2e-tests.sh
2022/07/04 14:20:30 Getting artifacts dir from prow
2022/07/04 14:20:30 Will use boskos to provision the GCP project
2022/07/04 14:20:30 Running kubetest2 with flags: ["gke" "--up" "--down" "--ignore-gcp-ssh-key=true" "--boskos-heartbeat-interval-seconds=60" "-v=1" "--cluster-version=latest" "--create-command=beta container clusters create --quiet --enable-autoscaling --min-nodes=1 --max-nodes=3 --scopes=cloud-platform --addons=NodeLocalDNS  --logging=NONE --monitoring=NONE" "--cluster-name=e2e-cls-wudddqjn" "--environment=prod" "--num-nodes=1" "--machine-type=e2-standard-4" "--network=e2e-network" "--image-type=cos_containerd" "--boskos-acquire-timeout-seconds=1200" "--region=us-east1,us-west1,us-central1" "--retryable-error-patterns='.*does not have enough resources available to fulfill.*,.*only \\d+ nodes out of \\d+ have registered; this is likely due to Nodes failing to start correctly.*,.*All cluster resources were brought up.+ but: component .+ from endpoint .+ is unhealthy.*'" "--test=exec" "--" "/home/prow/go/src/knative.dev/net-gateway-api/test/e2e-tests.sh" "--run-tests"]
I0704 14:20:30.168782     255 app.go:61] RunDir for this run: "/logs/artifacts/6e4cb39f-fba4-11ec-a513-1620173fdc34"
I0704 14:20:30.171785     255 app.go:120] ID for this run: "6e4cb39f-fba4-11ec-a513-1620173fdc34"
I0704 14:20:30.171812     255 up.go:344] explicit --cluster-name specified, ignoring --num-clusters
I0704 14:20:30.171888     255 common.go:69] No GCP projects provided, acquiring from Boskos [1] project/s
I0704 14:20:30.188595     255 common.go:92] Got project knative-boskos-41 from boskos
Updated property [core/project].
I0704 14:20:32.951749     255 network.go:169] Couldn't describe network "e2e-network", assuming it doesn't exist and creating it
Created [https://www.googleapis.com/compute/v1/projects/knative-boskos-41/global/networks/e2e-network].
NAME         SUBNET_MODE  BGP_ROUTING_MODE  IPV4_RANGE  GATEWAY_IPV4
e2e-network  AUTO         REGIONAL
```

Both these bits:
- `The connection to the server localhost:8080 was refused - did you specify the right host or port?`
- `KO_DOCKER_REPO: You must set 'KO_DOCKER_REPO'`

evankanderson commented:
Hmm -- I'm not sure about the first warning about localhost:8080; for the second, it seems like the environment variable is not set, but should be `kind.local` for the Kind tests. Not sure how it's set for other repos, though.

carlisia commented:
However it is set (`KO_DOCKER_REPO`) it works on `main` ¿ⓧ_ⓧﮌ

carlisia commented:
No `The connection to the server localhost:8080 was refused`, progress maybe 
![image](https://user-images.githubusercontent.com/16508/177441712-61580417-8897-4dea-bb24-87fc92b62c7e.png)

dprotaso commented on test/setup-and-deploy.sh:
```diff
@@ -0,0 +1,63 @@
+# This script includes common functions for setting up different Gateway API implementations.
+source "$(dirname $0)"/e2e-common.sh
+
+test_setup
```
the `hack` scripts already call this - you shouldn't need to

ie. see lifecycle: https://github.com/knative/hack#using-the-e2e-testssh-helper-script

dprotaso commented on test/setup-and-deploy.sh:
```diff
@@ -0,0 +1,63 @@
+    kubectl get pods --all-namespaces
+}
+
+function deploy_istio() {
```
You could move these functions into the `e2e-common.sh` ? Or have `e2e-common.sh` source this file (vs. sourcing multiple files in `*-test.sh`)

dprotaso commented on test/e2e-tests.sh:
```diff
@@ -14,13 +14,34 @@
+
 failed=0
 
+function test_setup() {
+    deploy_contour
+    deploy_istio
+}
```
This override has to be done before calling `initialize`

At this point the test is already running

carlisia commented:
- ✅ conformance tests
	- there's 1 failing but it's the same failure in `main`: `e2e tests (v1.23.6)`
	- `Error meeting response expectations: got unexpected status: 503, expected map[200:{}]`
	- considering parity with main a good bar for this PR :) (unless the fix is obvious)
- [ ] e2e tests
- [ ] integration-tests_net-gateway-api_main

carlisia commented:
Tests are passing now.

evankanderson commented:
(Adding this to the queue of reviews I need to do tonight)

evankanderson commented on pkg/reconciler/ingress/resources/httproute.go:
```diff
@@ -62,44 +62,44 @@ func MakeHTTPRoute(
-	gatewayRef := gatewayv1alpa2.ParentRef{
-		Namespace: namespacePtr(gatewayv1alpa2.Namespace(namespacedNameGateway.Namespace)),
-		Name:      gatewayv1alpa2.ObjectName(namespacedNameGateway.Name),
+	gatewayRef := gatewayv1alpha2.ParentReference{
```
Reviewing note: Combining the typo fix with the `s/ParentRef/ParentReference/` makes it harder to scan n' approve the file -- there are at least two different red/green patterns to look for.

evankanderson commented on test/e2e-library.sh:
```diff
@@ -0,0 +1,154 @@
+}
+
+# The tests below are run in a Kind cluster, therefore they need to have their ingressendpoint and cluster-suffix specified.
+function cluster_info() {
+    CONTROL_NAMESPACE=knative-serving
+    IPS=( $(kubectl get nodes -lkubernetes.io/hostname!=kind-control-plane -ojsonpath='{.items[*].status.addresses[?(@.type=="InternalIP")].address}') )
+    CLUSTER_SUFFIX=${CLUSTER_SUFFIX:-cluster.local}
+}
```
`cluster_info` here feels very similar to `export_cluster_info` in `e2e-common.sh`. Combine?

evankanderson replied:
(Feel free to make this a TODO if it feels too fragile to do right now)

evankanderson commented on test/e2e-library.sh:
```diff
@@ -0,0 +1,154 @@
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=1
+}
+
+function kind_e2e_istio() {
+    export GATEWAY_OVERRIDE=""
+    export GATEWAY_NAMESPACE_OVERRIDE=""
+
+    cluster_info
+
+    go test -race -count=1 -short -timeout=20m -tags=e2e ./test/conformance \
+        --enable-alpha --enable-beta \
+        --skip-tests="${ISTIO_UNSUPPORTED_E2E_TESTS}" \
+        --ingressendpoint="${IPS[0]}" \
+        --ingressClass=gateway-api.ingress.networking.knative.dev \
+        --cluster-suffix="$CLUSTER_SUFFIX"
+
+    # Give the controller time to sync with the rest of the system components.
+    sleep 30
+
+    echo ">> Scale up controller for HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=2
+
+    go test -count=1 -timeout=15m -failfast -parallel=1 -tags=e2e ./test/ha -spoofinterval="10ms" \
+        --enable-alpha --enable-beta \
+        --ingressendpoint="${IPS[0]}" \
+        --ingressClass=gateway-api.ingress.networking.knative.dev \
+        --cluster-suffix="${CLUSTER_SUFFIX}"
+
+    echo ">> Scale down after HA tests"
+    kubectl -n "${CONTROL_NAMESPACE}" scale deployment net-gateway-api-controller --replicas=1
+}
```
This feel similar-but-not-identical to both `kind_e2e_contour` and `e2e_istio`. Might be worth a TODO to deduplicate more _in a future PR_.

evankanderson commented on test/e2e-tests.sh:
```diff
@@ -14,27 +14,20 @@
+header "Running e2e tests with all available Gateway API vendors installed"
+e2e_istio
 
-success
```
We used to call `success` at the end of this script, and now we don't.

It looks like that function prints a success banner, but then also dumps some metrics from the kubernetes apiserver.

evankanderson reviewed (commented):
Looks fairly good -- I think only one or two of these are actionable.

/approve

(Will `/lgtm` if you decide that you want to handle followups as "quick win" subsequent changes.)

carlisia commented on .gitignore:
```diff
@@ -9,3 +9,5 @@
 # Temporary output of build tools
 bazel-*
 *.out
+
+tmp
```
Not mine. This tmp dir gets created during the code generation. Because it was taking so long/freezing on my machine and I had to kill that process many times, I noticed that depending on how far in the process the dir wouldn't have been cleaned up. 

I could have added that to my global gitignore but figured it will be useful in the case when interrupting the codegen before it runs to the end.

carlisia commented on DEVELOPMENT.md:
```diff
@@ -1,187 +1,77 @@
 This doc explains how to setup a development environment so you can get started
-[contributing](https://www.knative.dev/contributing/) to Knative
-`net-gateway-api`. Also take a look at:
+contributing to Knative `net-gateway-api`.
```
I moved it to a bullet the very next line, next to another link of similar kind.

carlisia commented:
I could compile your comments into a todo list and have a followup PR today. I'm concerned of holding up this PR even longer. Am open to adding anything that is a must include in this PR, just having a hard time distinguishing if anything is. At any rate, thank you so much for the throughout review.

evankanderson commented:
/lgtm

Notes from slack:

I think the comments on DEVELOPMENT.md, [such as this one](https://github.com/knative-sandbox/net-gateway-api/pull/315#discussion_r917225763) were the ones I was most concerned about fixing, since they might make it harder for someone else to join in.

I’m also wondering about putting `100-gateway-api.yaml` into our release, but it looks like that may be long-standing, so I’d TODO it.

Can you also open an issue for [the addition of `ReferencePolicies CRD](https://github.com/knative-sandbox/net-gateway-api/pull/315#discussion_r917226606) to the `100-gateway-api.yaml` which is marked as temporary in `update-codegen.sh`?
